{
 "cells": [
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-03T14:13:24.905489Z",
     "start_time": "2024-07-03T14:13:24.176681Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# some initial imports\n",
    "\n",
    "import jax.numpy as jnp\n",
    "import jax.lax"
   ],
   "id": "1d81059b95845310",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Using the JAX-powered resource-aware cell modelling package _rc_e_coli_jax_ for gene circuit simulations\n",
    "\n",
    "This Jupyter notebook provides a step-by-step guide for simulating how a gene circuit behaves in the context of the host cell and other competing synthetic genes. Underlying these simulations is a resource-aware coarse-grained cell model published in [Sechkar et al., 2024](https://doi.org/10.1038/s41467-024-46410-9), which here has been implemented in JAX to enable efficient parallelised simulations on the GPU.\n",
    "\n",
    "As an example case, we consider the case of a gene _ta_ that encodes a transcription activation factor which, upon chemical modulation, promotes the expression of the output gene _x_. In addition to these and the host cell's own genes, we consider another disturbing synthetic gene _dist_ that competes with _ta_ and _x_ for the host's gene expression resources. Each gene has two modelled variables associated with it: its mRNA concentration in the cell _m_ and its protein level _p_. The resultant circuit is depicted in the figure below (adapted from [Sechkar et al., 2024](https://doi.org/10.1038/s41467-024-46410-9)).\n",
    "\n",
    "<div>\n",
    "<img src=\"example.png\" width=\"750\"/>\n",
    "</div>"
   ],
   "id": "28a05258b9c48a9a"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Circuit model definition\n",
    "\n",
    "We start by defining the functions that allow to model the circuit of interest -- namely, the initialiser _initialise()_, the gene regulation function _F_calc()_, the deterministic ordinary differential equation (ODE) model function _ode()_, and - if needed - a reaction propensity function _v()_ for stochastic simulations. \n",
    "\n",
    "In the file _one_constit.py_ they are specified for simulating a single constituive gene present in the cell. For any new gene circuit, one can copy and paste this file's contents into a new file - then modify the circuit-specific code fragments, which have all been highighted as follows:\n",
    "\n",
    "````{python}\n",
    "    # -------- SPECIFY [X] FROM HERE...\n",
    "    # -------- ...TO HERE\n",
    "````"
   ],
   "id": "237e81acf96a3d34"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Circuit initialisation\n",
    "In our case this means that in the initialiser function, the genes specified are _ta_, _x_ and _dist_. For out case study, we assume that their DNA concentrations $c$ and promoter strengths $\\alpha$ are not necessarily the deafult values, so we re-specify them below. We additionally set the parameters of the Hill functions that describe the chemical inducer's binding to the transcription activation factors and the subsequent binding between the output gene's DNA and the inducer-bound transcription activation factor."
   ],
   "id": "c72c40fd6aa3cc5e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-03T14:13:24.911396Z",
     "start_time": "2024-07-03T14:13:24.906439Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def initialise():\n",
    "    # -------- SPECIFY CIRCUIT COMPONENTS FROM HERE...\n",
    "    genes = ['ta','x','dist']  # names of genes in the circuit\n",
    "    miscs = []  # names of miscellaneous species involved in the circuit (e.g. metabolites)\n",
    "    # -------- ...TO HERE\n",
    "\n",
    "    # for convenience, one can refer to the species' concs. by names instead of positions in x\n",
    "    # e.g. x[name2pos['m_xtra']] will return the concentration of mRNA of the gene 'xtra'\n",
    "    name2pos = {}\n",
    "    for i in range(0, len(genes)):\n",
    "        name2pos['m_' + genes[i]] = 8 + i  # mRNA\n",
    "        name2pos['p_' + genes[i]] = 8 + len(genes) + i  # protein\n",
    "    for i in range(0, len(miscs)):\n",
    "        name2pos[miscs[i]] = 8 + len(genes) * 2 + i  # miscellaneous species\n",
    "    for i in range(0, len(genes)):\n",
    "        name2pos['k_' + genes[i]] =  i  # effective mRNA-ribosome dissociation constants (in k_het, not x!!!)\n",
    "    for i in range(0, len(genes)):\n",
    "        name2pos['F_' + genes[i]] =  i  # transcription regulation functions (in F, not x!!!)\n",
    "    for i in range(0, len(genes)):\n",
    "        name2pos['mscale_' + genes[i]] =  i  # mRNA count scaling factors (in mRNA_count_scales, not x!!!)\n",
    "\n",
    "    # default gene parameters to be imported into the main model's parameter dictionary\n",
    "    default_par = {}\n",
    "    for gene in genes: # gene parameters\n",
    "        default_par['func_' + gene] = 1.0  # gene functionality - 1 if working, 0 if mutated\n",
    "        default_par['c_' + gene] = 1.0  # copy no. (nM)\n",
    "        default_par['a_' + gene] = 100.0  # promoter strength (unitless)\n",
    "        default_par['b_' + gene] = 6.0  # mRNA decay rate (/h)\n",
    "        default_par['k+_' + gene] = 60.0  # ribosome binding rate (/h/nM)\n",
    "        default_par['k-_' + gene] = 60.0  # ribosome unbinding rate (/h)\n",
    "        default_par['n_' + gene] = 300.0  # protein length (aa)\n",
    "        default_par['d_' + gene] = 0.0  # rate of active protein degradation - zero by default (/h)\n",
    "\n",
    "    # default initial conditions\n",
    "    default_init_conds = {}\n",
    "    for gene in genes:\n",
    "        default_init_conds['m_' + gene] = 0\n",
    "        default_init_conds['p_' + gene] = 0\n",
    "    for misc in miscs:\n",
    "        default_init_conds[misc] = 0\n",
    "\n",
    "    # -------- DEFAULT VALUES OF CIRCUIT-SPECIFIC PARAMETERS CAN BE SPECIFIED FROM HERE...\n",
    "    \n",
    "    # gene concentrations (nM) and promoter strengths (unitless)\n",
    "    default_par['c_ta']=100\n",
    "    default_par['c_x']=100\n",
    "    default_par['a_ta']=50\n",
    "    default_par['a_x']=50\n",
    "    \n",
    "    # binging between the chemical inducer and the transcription activation factor\n",
    "    default_par['K_ta-f']=1000  # Half-saturation constant (nM)\n",
    "    \n",
    "    # binding between the output gene's DNA and the inducer-bound transcription activation factor\n",
    "    default_par['K_dna(x)-taf']=700  # Half-saturation constant (nM)\n",
    "    default_par['eta_dna(x)-taf']=2 # Hill coefficient/cooperativity of binding\n",
    "    default_par['baseline']=0.1 # baseline output gene promoter activity in abscence of binding\n",
    "    \n",
    "    # time of the inducer's addition to the medium (h)\n",
    "    default_par['t_add']=25\n",
    "    default_par['f_added']=1000  # added inducer concentration (nM)\n",
    "    # -------- ...TO HERE\n",
    "\n",
    "    # default palette and dashes for plotting (5 genes + misc. species max)\n",
    "    default_palette = [\"#0072BD\", \"#D95319\", \"#4DBEEE\", \"#A2142F\", \"#FF00FF\"]\n",
    "    default_dash = ['solid']\n",
    "    # match default palette to genes and miscellaneous species, looping over the five colours we defined\n",
    "    circuit_styles = {'colours': {}, 'dashes': {}}  # initialise dictionary\n",
    "    for i in range(0, len(genes)):\n",
    "        circuit_styles['colours'][genes[i]] = default_palette[i % len(default_palette)]\n",
    "        circuit_styles['dashes'][genes[i]] = default_dash[i % len(default_dash)]\n",
    "    for i in range(len(genes), len(genes) + len(miscs)):\n",
    "        circuit_styles['colours'][miscs[i - len(genes)]] = default_palette[i % len(default_palette)]\n",
    "        circuit_styles['dashes'][miscs[i - len(genes)]] = default_dash[i % len(default_dash)]\n",
    "\n",
    "    # --------  YOU CAN RE-SPECIFY COLOURS FOR PLOTTING FROM HERE...\n",
    "    # -------- ...TO HERE\n",
    "\n",
    "    return default_par, default_init_conds, genes, miscs, name2pos, circuit_styles"
   ],
   "id": "4d5ba1e025d6e1fe",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Gene transcription regulation\n",
    "\n",
    "We now proceed to define the gene transcription regulation function _F_calc()_.\n",
    "\n",
    "As mentioned above, the genes _ta_ and _dist_ are constitutive, so the value of the gene transcription regulation function _F_ will be 1 at all times. For the gene _ta_, regulation happens as follows. First, the inducer with the concentration $f$ binds the transcription activation protein $p_{ta}$, hence the concentration of the active inducer-bound factor being given by a Hill function:\n",
    "$$p_{ta}^{act}=p_{ta} \\frac{f}{f+K_{ta-f}}$$\n",
    "Second, the active transcription activation factor binds the output gene's DNA, hence the output gene's promoter activity being given by another Hill function:\n",
    "$$F_{x}=baseline+\\frac{(p_{ta}^{act})^{\\eta_{dna(x)-taf}}}{(p_{ta}^{act})^{\\eta_{dna(x)-taf}} + (K_{dna(x)-taf})^{\\eta_{dna(x)-taf}}$$\n",
    "\n",
    "The inducer may not be originally present in the culture medium, but rather than added to it at some time $t_{add}$ as a pulse input (see below). This can be recreated by using the _jax.lax.select()_ function in _F_calc()_.\n",
    "$$f(t)=\\begin{cases}0 & \\text{ if } t<t_{add} \\\\ f_{added} & \\text{ if } t\\geq t_{add}\\end{cases}$$\n"
   ],
   "id": "16bfad32b9842f20"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-07-03T14:13:24.918138Z",
     "start_time": "2024-07-03T14:13:24.912451Z"
    }
   },
   "source": [
    "def F_calc(t ,x, par, name2pos):\n",
    "    # --------  SPECIFY THE TRANSCRIPTIOPN REGULATION FUNCTION FROM HERE...\n",
    "    F_ta = 1 # ta gene is constitutive\n",
    "    F_dist = 1 # dist gene is constitutive\n",
    "    \n",
    "    # get the time-dependent inducer concentration\n",
    "    f = jax.lax.select(t<par['t_add'], 0, par['f_added'])\n",
    "    \n",
    "    # get the concentration of the transcription activation factor using the 'variable name-position in state vector' mapper\n",
    "    p_ta = x[name2pos['p_ta']] \n",
    "    \n",
    "    # binding between the chemical inducer and the transcription activation factor\n",
    "    p_ta_act = p_ta * f/(f+par['K_ta-f'])\n",
    "    \n",
    "    # binding between the output gene's DNA and the inducer-bound transcription activation factor\n",
    "    F_x = par['baseline']+(p_ta_act**par['eta_dna(x)-taf'])/(p_ta_act**par['eta_dna(x)-taf']+par['K_dna(x)-taf']**par['eta_dna(x)-taf'])\n",
    "    \n",
    "    # returning the regulation function values in the same order as we specified the genes in the 'initialise()' function\n",
    "    return jnp.array([F_ta, F_x, F_dist])\n",
    "    # -------- ...TO HERE"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Deterministic ODE model definition\n",
    "\n",
    "We now define the deterministic ODE model function _ode()_, taken from Equations (S128) to (S133) of the Supplementary Information to [Sechkar et al., 2024](https://doi.org/10.1038/s41467-024-46410-9):\n",
    "\n",
    "\\begin{align}\n",
    "    \\dot{m_{ta}} &= F_{ta} c_{ta} \\alpha_{ta} \\lambda(\\epsilon,B) - (\\beta_{ta} + \\lambda(\\epsilon,B))m_{ta}\n",
    "    \\\\\n",
    "    \\dot{m_{x}} &= F_{x}(f,p_{ta}) \\cdot c_{x} \\alpha_{x} \\lambda(\\epsilon,B) - (\\beta_{x} + \\lambda(\\epsilon,B))m_{x}\n",
    "    \\\\\n",
    "    \\dot{m}_{dist} &= F_{dist} c_{dist} \\alpha_{dist} \\lambda(\\epsilon,B) - (\\beta_{dist} + \\lambda(\\epsilon,B))m_{dist}\n",
    "    \\\\  \n",
    "    \\dot{p_{ta}} &= \\frac{\\epsilon(t^c)}{n_{ta}} \\cdot \n",
    "    \\frac{m_{ta} / k_{ta}}{D} R - (\\delta_{ta} + \\lambda(\\epsilon,B)) \\cdot p_{ta}\n",
    "    \\\\\n",
    "    \\dot{p_{x}} &= \\frac{\\epsilon(t^c)}{n_{x}} \\cdot \n",
    "    \\frac{m_{x} / k_{x}}{D} R - (\\delta_{x} + \\lambda(\\epsilon,B)) \\cdot p_{x}\n",
    "    \\\\\n",
    "    \\dot{p}_{dist} &= \\frac{\\epsilon(t^c)}{n_{dist}} \\cdot \n",
    "    \\frac{m_{dist} / k_{dist}}{D} R - (\\delta_{dist} + \\lambda(\\epsilon,B)) \\cdot p_{dist}\n",
    "\\end{align}\n",
    "\n",
    "Here, $\\lambda$ is the cell's growth rate, $R$ is the concentration of ribosomes in the cell, $\\epsilon$ is the translation elongation rate, and $D$ is the 'resource competition denomionator' capturing the extent of competition for gene expression resources (i.e. ribosomes) in the cell. For gene $i$, $F_i$ its transcription regulation function (implemented in _F_calc()_ as discussed above), $c_i$ is its DNA concentration, $\\alpha_i$ is its promoter strength, $\\beta_i$ is its mRNA degradation rate, $\\delta_i$ is the protein degradation rate, $n_i$ is the number of amino acids in the encoded protein, and $k_i$ is the effective mRNA-ribosome dissociation constant. A more detailed discussion of the model equations can be found in [Sechkar et al., 2024](https://doi.org/10.1038/s41467-024-46410-9).\n",
    "\n",
    "In our JAX implementation _ode()_, each of the ODEs above is simply typed into the corresponding entry in the array dxdt, where the order of the genes is the same as specified in _initialise()_ and all mRNA level ODEs come before those for the protein concentrations. Once again, the variable with a given name can be retrieved from the state vector using the name2pos mapper, e.g. _x[name2pos['m_ta']]_ will return the value of the mRNA concentration of the gene _ta_."
   ],
   "id": "fa736a20b77a47ca"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "def ode(F_calc,     # calculating the transcription regulation functions\n",
    "            t,  x,  # time, cell state, external inputs\n",
    "            e, l, # translation elongation rate, growth rate\n",
    "            R, # ribosome count in the cell, resource\n",
    "            k_het, D, # effective mRNA-ribosome dissociation constants for synthetic genes, resource competition denominator\n",
    "            par,  # system parameters\n",
    "            name2pos  # name to position decoder\n",
    "            ):\n",
    "    # GET REGULATORY FUNCTION VALUES\n",
    "    F = F_calc(t, x, par, name2pos)\n",
    "\n",
    "    # --------  SPECIFY THE ODEs FROM HERE...\n",
    "    return [# mRNAs\n",
    "            F[name2pos['F_ta']] * par['c_ta'] * par['a_ta'] * e - (par['b_ta'] + l) * x[name2pos['m_ta']],  # m_ta\n",
    "            F[name2pos['F_x']] * par['c_x'] * par['a_x'] * e - (par['b_x'] + l) * x[name2pos['m_x']],  # m_x\n",
    "            F[name2pos['F_dist']] * par['c_dist'] * par['a_dist'] * e - (par['b_dist'] + l) * x[name2pos['m_dist']],  # m_dist\n",
    "            # proteins\n",
    "            (e / par['n_ta']) * (x[name2pos['m_ta']] / k_het[name2pos['k_ta']] / D) * R - (l + par['d_ta']) * x[name2pos['p_ta']],  # p_ta\n",
    "            (e / par['n_x']) * (x[name2pos['m_x']] / k_het[name2pos['k_x']] / D) * R - (l + par['d_x']) * x[name2pos['p_x']],  # p_x\n",
    "            (e / par['n_dist']) * (x[name2pos['m_dist']] / k_het[name2pos['k_dist']] / D) * R - (l + par['d_dist']) * x[name2pos['p_dist']]  # p_dist\n",
    "            ]\n",
    "    # -------- ...TO HERE"
   ],
   "id": "7d9420de71fea337"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Stochastic model definition\n",
    "\n",
    "To account for the stochasticity of gene expression, our package also allows to perform hybrid simulations of gene circuit performance, where the host cell variables are still treated deterministically (since they are coarse-grained variables representing the mean dynamics of many different variables, whose fluctuations are averaged out)and the synthetic gene variables are treated stochastically. \n",
    "\n",
    "Stochastic simulation of synthetic gene expression requires to define the reaction propensity function _v()_. This can be done simply by putting each ODE term, in order of appearance in the ODE array above, into a separate entry in the propensity vector, with a few caveats. \n",
    "\n",
    "First, due to the way that the translation of a single mRNA by several ribosomes is modelled, all terms involving mRNAs must be scaled by a factor found in the _mRNA_count_scales_ vector, which can be accessed using the same decoder _name2pos_. \n",
    "\n",
    "Second, the model allows to consider mRNA removal due to antibiotic action, which we do not do in this case. However, to maintain the correct order of the propensities, we include an additional term, set to zero, for all mRNAs. Should it be needed, the details of how to model antibiotic action can be found in the Supplementary Information to [Sechkar et al., 2024](https://doi.org/10.1038/s41467-024-46410-9)."
   ],
   "id": "4f0cb866c8db5507"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "def v(F_calc,     # calculating the transcription regulation functions\n",
    "            t,  x,  # time, cell state, external inputs\n",
    "            e, l, # translation elongation rate, growth rate\n",
    "            R, # ribosome count in the cell, resource\n",
    "            k_het, D, # effective mRNA-ribosome dissociation constants for synthetic genes, resource competition denominator\n",
    "            mRNA_count_scales, # scaling factors for mRNA counts\n",
    "            par,  # system parameters\n",
    "            name2pos\n",
    "            ):\n",
    "    # GET REGULATORY FUNCTION VALUES\n",
    "    F = F_calc(t, x, par, name2pos)\n",
    "\n",
    "    # --------  SPECIFY THE PROPENSITIES FROM HERE...\n",
    "    return [\n",
    "            # synthesis, degradation, dilution of ta gene mRNA - note the scaling factor added\n",
    "            l * F[name2pos['F_ta']] * par['c_ta'] * par['a_ta'] / mRNA_count_scales[name2pos['mscale_ta']],\n",
    "            par['b_ta'] * x[name2pos['m_ta']] / mRNA_count_scales[name2pos['mscale_ta']],\n",
    "            l * x[name2pos['m_ta']] / mRNA_count_scales[name2pos['mscale_ta']],\n",
    "            # mRNA removal due to chloramphenicol action - set to zero\n",
    "            0,\n",
    "            # synthesis, degradation, dilution of x gene mRNA - note the scaling factor added\n",
    "            l * F[name2pos['F_x']] * par['c_x'] * par['a_x'] / mRNA_count_scales[name2pos['mscale_x']],\n",
    "            par['b_x'] * x[name2pos['m_x']] / mRNA_count_scales[name2pos['mscale_x']],\n",
    "            l * x[name2pos['m_x']] / mRNA_count_scales[name2pos['mscale_x']],\n",
    "            # mRNA removal due to chloramphenicol action - set to zero\n",
    "            0,\n",
    "            # synthesis, degradation, dilution of dist gene mRNA - note the scaling factor added\n",
    "            l * F[name2pos['F_dist']] * par['c_dist'] * par['a_dist'] / mRNA_count_scales[name2pos['mscale_dist']],\n",
    "            par['b_dist'] * x[name2pos['m_dist']] / mRNA_count_scales[name2pos['mscale_dist']],\n",
    "            l * x[name2pos['m_dist']] / mRNA_count_scales[name2pos['mscale_dist']],\n",
    "            # mRNA removal due to chloramphenicol action - set to zero\n",
    "            0,\n",
    "            # synthesis, degradation, dilution of ta gene protein\n",
    "            (e / par['n_ta']) * (x[name2pos['m_ta']] / k_het[name2pos['k_ta']] / D) * R,\n",
    "            par['d_ta'] * x[name2pos['p_ta']],\n",
    "            l * x[name2pos['p_ta']],\n",
    "            # synthesis, degradation, dilution of x gene protein\n",
    "            (e / par['n_x']) * (x[name2pos['m_x']] / k_het[name2pos['k_x']] / D) * R,\n",
    "            par['d_x'] * x[name2pos['p_x']],\n",
    "            l * x[name2pos['p_x']],\n",
    "            # synthesis, degradation, dilution of dist gene protein\n",
    "            (e / par['n_dist']) * (x[name2pos['m_dist']] / k_het[name2pos['k_dist']] / D) * R,\n",
    "            par['d_dist'] * x[name2pos['p_dist']],\n",
    "            l * x[name2pos['p_dist']]\n",
    "    ]\n",
    "    # -------- ...TO HERE"
   ],
   "id": "155095fe0b65da98"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Gene circuit simulation\n",
    "\n",
    "Now that the circuit has been defined, we can proceed to simulate its behaviour."
   ],
   "id": "74f2dcc0cc6fae54"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
